### PR TITLE
Array pattern

### LINKED ISSUE #5: Function parameters with defaults not shortened
Currently code like the following will cause an error:

```javascript
function test (myParam = 1) { }
```

PR #3 will stop the error but we should be able to still shorten these names.

### PR DESCRIPTION
**Pull request by imranolas**
This adds mangling for destructured arrays. There's scope to handle rest params arguments too.

Fixes #5

### CODE REVIEW

File: spec/fixtures/shorten-array-pattern-variables/expected.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+const [a, b, c] = [1, 2, 3];
+
+const myFunc = ([d, e, f]) => null;
```
Comment (sheepsteak) [inline spec/fixtures/shorten-array-pattern-variables/expected.js]: One things that bothers me is that technically these function parameters could also be `[a, b, c]` as the other `[a, b, c]` isn't used anywhere in the function. But I guess the function's scope can see them so they're considered "taken".

**Comment by sheepsteak** [comment]
👍 This is great. 

It would be really good to get rest params done too whilst you're here. :trollface: 

Looks like that is a `RestElement` in the AST so should be fairly simple:

screen shot 2016-05-17 at 23 57 38

**Comment by imranolas** [comment]
Rest elements, default function arguments and destructured array arguments are now mangled. 

Fixes #5.

File: spec/fixtures/shorten-array-pattern-variables/actual.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+const [variableA, variableB = 2, ...variableC] = [1, 2, 3, 4];
```
Comment (sheepsteak) [inline spec/fixtures/shorten-array-pattern-variables/actual.js]: I didn't even realise you could give an array element a default value when destructured!

File: src/index.js [hunk 3]
```diff
@@ -24,6 +24,25 @@ function generateVariableName(scope) {
...
+        scope.rename(element.argument.name, newName);
+      } else if (t.isAssignmentPattern(element)) {
+        renameAssignmentIdentifier(scope, element);
+      }
+    });
+  }
+
+  function renameAssignmentIdentifier(scope, node) {
+    const newName = generateVariableName(scope);
+    scope.rename(node.left.name, newName);
```
Comment (sheepsteak) [inline src/index.js]: For these (and the ones above) we've not got the check to see whether it's worth getting a new variable (`length > 1`) and whether the new one is actually shorter than the current. Maybe we can refactor it into a function?

Comment (imranolas) [inline src/index.js]: Good catch. I've refactored so it's now all done in common place.

**Comment by sheepsteak** [comment]
👍

**Pull request merged**